### ISSUE TITLE
Object references are cached in NewtonsoftJsonOutputFormatter

### ISSUE THREAD

**Issue by weichch** [dotnet/aspnetcore#20672]
### Describe the bug
I guess this is more of a question than a bug or a feature request. The [original SO post](https://stackoverflow.com/questions/59832635/is-there-a-way-to-reset-id-in-json-response-for-each-http-request-jsonserializ) for cross reference.

I am using output formatters from `Newtonsoft.Json` with serializer settings:

```
opt.SerializerSettings.PreserveReferencesHandling = PreserveReferencesHandling.Objects;
```

When I use `StatusCode` method in controller actions to return object results, I noticed that the `$id` is cached. And won't reset for different requests.

For example, for request 1, I get this:
```
{
    "$id": "1"
}
```
For request 2, I get this:
```
{
    "$id": "2"
}
```

I dug up the [source code here](https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonOutputFormatter.cs#L26), and realized that the JSON serializer is cached for performance reasons.

So I have two questions here:

Seems like the [DefaultReferenceResolver](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Serialization/DefaultReferenceResolver.cs) in Newtonsoft.Json uses a dictionary to keep object reference mappings. I can't see anywhere the dictionary gets cleared / item removed.

**Would this cause any memory issues?**

I know I can set to use my own `IReferenceResolver`, but again the resolver instance is also cached. So to reset the counter for each request, I need to go a long way, possibly need to figure out / activate some storage in the request scope. 

**Is there a simpler way I can reset the counter for each request?**

### To Reproduce

Startup.cs
```
services.AddControllers()
    .AddNewtonsoftJson(opt =>
    {
        opt.SerializerSettings.PreserveReferencesHandling = PreserveReferencesHandling.Objects;
        // This function here gets called only once, so it is like static / singleton
        // opt.SerializerSettings.ReferenceResolverProvider = () => new MyReferenceResolver();
    });
```

Controller.cs
```
[HttpPost]
public IActionResult Create()
{
    return StatusCode(200, new { });
}
```

### Further technical details
- ASP.NET Core version
```
Microsoft.AspNetCore.Mvc.NewtonsoftJson 3.1.3
```
- Include the output of `dotnet --info`
```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.201
 Commit:    b1768b4ae7
```
- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version: VS2019 16.5.3

**Comment by davidfowl** [comment]
cc @JamesNK

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @weichch.
We will look into this soon and get back to you with our findings.

**Comment by pranavkm** [comment]
Thanks for the report @weichch. We've addressed this as part of 2.1.18 and 3.1.4 release.
